### DISCUSSION TITLE
Trouble removing a module from Pyodide's Python, to make it smaller

### DISCUSSION THREAD

**Discussion by davidfstr** [pyodide/pyodide#5141] [Q&A] [answered]
I am attempting to remove a module from CPython inside Pyodide to make it smaller.

I build Pyodide inside a Docker container using the Pyodide Docker image, on an ARM Mac:

```
$ git clone git@github.com:pyodide/pyodide.git
$ git checkout stable  # 0.26.3
$ cd pyodide
$ ./run_docker
$$ make
$$ ls -l dist/pyodide.asm.wasm* dist/python_stdlib.zip*
-rwxr-xr-x@ 1 dfoster  staff  10086131 Oct 25 14:07 dist/pyodide.asm.wasm
-rw-r--r--  1 dfoster  staff   2341804 Oct 25 14:07 dist/python_stdlib.zip
```

I try to modify cpython/Setup.local to remove a module:

```
$$ echo "*disabled*" >> cpython/Setup.local
$$ echo "_asyncio" >> cpython/Setup.local
```

But `make` does not detect the changes, resulting in identical output in `./dist`:

```
$$ make
$$ ls -l dist/pyodide.asm.wasm* dist/python_stdlib.zip*
-rwxr-xr-x@ 1 dfoster  staff  10086131 Oct 25 14:07 dist/pyodide.asm.wasm
-rw-r--r--  1 dfoster  staff   2341804 Oct 25 14:07 dist/python_stdlib.zip
```

I try to force the "Building cpython..." rule inside the root Makefile to rerun:

```
$$ make -C cpython
...
Traceback (most recent call last):
  File "/src/cpython/build/Python-3.12.1/./Tools/wasm/wasm_assets.py", line 257, in <module>
    main()
  File "/src/cpython/build/Python-3.12.1/./Tools/wasm/wasm_assets.py", line 228, in main
    args.sysconfig_data = get_sysconfigdata(args)
                          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/src/cpython/build/Python-3.12.1/./Tools/wasm/wasm_assets.py", line 128, in get_sysconfigdata
    raise ValueError(
ValueError: ("Invalid sysconfig data name '_sysconfigdata__linux_x86_64-linux-gnu'.", ('_sysconfigdata__emscripten_wasm32-emscripten', '_sysconfigdata__emscripten_wasm32-emscripten', '_sysconfigdata__wasi_wasm32-wasi', '_sysconfigdata__wasi_wasm64-wasi'))
```

Some questions:
1. Any tips for fixing the root Makefile to actually detect changes to `cpython/Setup.local`?
    * It looks like I should be able to rebuild with `make -C cpython`, but that shows a build error in wasm_assets.py.
2. Any insight as to why wasm_assets.py is having an error?
    * It looks like a "linux" environment is being detected when an "emscripten" or "wasi" environment was expected.

**Comment by hoodmane** [comment]
I've seen this error but only quite recently. I'm not really sure what changed to make it happen. In any case, try:
```
make -C cpython rebuild
```

**Reply by davidfstr** [reply]
The extra `rebuild` part of `make -C cpython rebuild` doesn't eliminate the wasm_assets.py error, unfortunately.

---

Debugging further, wasm_assets.py appears to be called by cpython/build/Python-3.12.1/Makefile with the line:
* `$(PYTHON_FOR_BUILD) $(srcdir)/Tools/wasm/wasm_assets.py ...`

where `$(PYTHON_FOR_BUILD)` is a version of python3.12 with a bunch of environment variables set:
```
PYTHON_FOR_BUILD=_PYTHON_PROJECT_BASE=$(abs_builddir) _PYTHON_HOST_PLATFORM=$(_PYTHON_HOST_PLATFORM) PYTHONPATH=$(shell test -f pybuilddir.txt && echo $(abs_builddir)/`cat pybuilddir.txt`:)$(srcdir)/Lib _PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata_$(ABIFLAGS)_$(MACHDEP)_$(MULTIARCH) /usr/local/bin/python3.12
```

However I don't actually see those environment variables set in the invocation of python3.12 that make prints:

```
/usr/local/bin/python3.12 ./Tools/wasm/wasm_assets.py \
    --buildroot . --prefix /src/cpython/installs/python-3.12.1
```

I don't know if make normally omits the environment variables from the command line, or if the lack of environment variables on the command line indicates a problem.

---

I *think* I've gotten out of this error state with a full rebuild (using `make clean-python` followed by `make`) but that takes over an hour to run on my computer. And it does nothing to prevent the error state from reappearing in the future.

**Reply by davidfstr** [reply]
Interestingly, directly rerunning the Make rule that called `wasm_assets.py` WAS effective:

```
$ make -C cpython/build/Python-3.12.1 wasm_stdlib
```

And running the rule directly DID show all the environment variables in the make output:

```
_PYTHON_PROJECT_BASE=/src/cpython/build/Python-3.12.1 _PYTHON_HOST_PLATFORM=emscripten-wasm32 PYTHONPATH=/src/cpython/build/Python-3.12.1/build/lib.linux-x86_64-3.12:./Lib _PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata__emscripten_wasm32-emscripten /usr/local/bin/python3.12 ./Tools/wasm/wasm_assets.py \
    --buildroot . --prefix /src/cpython/installs/python-3.12.1
```

---

But afterward trying to run the entire enclosing Makefile had its own error:

```
$ make -C cpython/build/Python-3.12.1
/src/emsdk/emsdk/upstream/emscripten/emcc   -sALLOW_MEMORY_GROWTH -sTOTAL_MEMORY=20971520 -sWASM_BIGINT -sFORCE_FILESYSTEM -lidbfs.js ...
wasm-ld: error: Modules/_ctypes/_ctypes.o: undefined symbol: ffi_type_pointer
(error repeated 13 times)
wasm-ld: error: Modules/_ctypes/callbacks.o: undefined symbol: ffi_closure_alloc
wasm-ld: error: Modules/_ctypes/callbacks.o: undefined symbol: ffi_type_void
wasm-ld: error: Modules/_ctypes/callbacks.o: undefined symbol: ffi_prep_cif
wasm-ld: error: Modules/_ctypes/callbacks.o: undefined symbol: ffi_prep_closure_loc
wasm-ld: error: Modules/_ctypes/callbacks.o: undefined symbol: ffi_type_void
wasm-ld: error: Modules/_ctypes/callbacks.o: undefined symbol: ffi_closure_free
wasm-ld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
emcc: error: '/src/emsdk/emsdk/upstream/bin/wasm-ld -o python.wasm ...' failed (returned 1)
```

I'm beginning to think the build system is haunted 👻

**Reply by hoodmane** [reply]
We don't even use the output of `wasm_assets.py`...

**Reply by hoodmane** [reply]
> I'm beginning to think the build system is haunted 👻

This problem has appeared rather inconsistently for me too.

**Reply by hoodmane** [reply]
I will try to investigate when I get a chance.

**Comment by ryanking13** [comment]
What happens if you don't use the docker image? We have some issue using docker image in ARM machines (https://github.com/pyodide/pyodide/issues/1386)

**Reply by davidfstr** [reply]
The last time I tried the build process "MacOS with Conda" documented [here](https://pyodide.org/en/stable/development/building-from-sources.html#using-make) I remember it not appearing to work, due to some error message. However when I try it again this morning it DOES seem to work. 🎉 

<details><summary>Successful build on ARM macOS with Conda</summary>

```
$ git clone git@github.com:pyodide/pyodide.git pyodide3-macos
$ cd pyodide3-macos
$ git checkout stable  # 0.26.3
$ xcode-select --install
$ conda env create -f environment.yml
$ conda activate pyodide-env
$ pip install -r requirements.txt
$ make
$ dist/python
Python 3.12.1 (main, Oct 30 2024, 11:10:51) [Clang 19.0.0git (https:/github.com/llvm/llvm-project 0a8cd1ed1f4f35905df318015b on emscripten
Type "help", "copyright", "credits" or "license" for more information.
>>> sum([1,2,3])
6
>>> ^D
```

</details> 

---

**Edit:** See next message in thread. Despite the initial build working, the incremental build still fails.

~~So I should be unblocked for now.~~

~~**I'll just consider the "Build on ARM macOS using the Dockerfile" to be an unsupported scenario.**~~

Once I'm past the prototyping stage in my project to slim down Pyodide, and ready to build something for production:
* I'll probably switch to the "Build on x86_64 Linux using the Dockerfile" process (running inside a virtual machine) since that seems to be the only configuration officially tested in CI.

**Reply by davidfstr** [reply]
Ah I spoke too soon. I *did* get the "MacOS with Conda" build process to actually work, but I still see both of the original problems:

1. `make` does not notice changes made to `cpython/Setup.local`
2. `make -C cpython` fails with an error from wasm_assets.py not getting the correct sysconfig information, although now it sees `_sysconfigdata__darwin_darwin` instead of `_sysconfigdata__linux_x86_64-linux-gnu`:

```
$ make -C cpython
...
/Users/dfoster/miniforge3/envs/pyodide-env/bin/python3.12 ./Tools/wasm/wasm_assets.py \
    --buildroot . --prefix /Users/dfoster/Documents/TechSmart/Repositories/pyodide3-macos/cpython/installs/python-3.12.1
Traceback (most recent call last):
  File "/Users/dfoster/Documents/TechSmart/Repositories/pyodide3-macos/cpython/build/Python-3.12.1/./Tools/wasm/wasm_assets.py", line 257, in <module>
    main()
  File "/Users/dfoster/Documents/TechSmart/Repositories/pyodide3-macos/cpython/build/Python-3.12.1/./Tools/wasm/wasm_assets.py", line 228, in main
    args.sysconfig_data = get_sysconfigdata(args)
                          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dfoster/Documents/TechSmart/Repositories/pyodide3-macos/cpython/build/Python-3.12.1/./Tools/wasm/wasm_assets.py", line 128, in get_sysconfigdata
    raise ValueError(
ValueError: ("Invalid sysconfig data name '_sysconfigdata__darwin_darwin'.", ('_sysconfigdata__emscripten_wasm32-emscripten', '_sysconfigdata__emscripten_wasm32-emscripten', '_sysconfigdata__wasi_wasm32-wasi', '_sysconfigdata__wasi_wasm64-wasi'))
emmake: error: 'make -C /Users/dfoster/Documents/TechSmart/Repositories/pyodide3-macos/cpython/build/Python-3.12.1 PYTHON_FOR_BUILD=/Users/dfoster/miniforge3/envs/pyodide-env/bin/python3.12 CROSS_COMPILE=yes inclinstall libinstall libpython3.12.a -j 3' failed (returned 2)
```

So I suspect there still some kind of bug in the build system that is not OS-specific.

**Reply by davidfstr** [reply, marked as answer]
Now that I have Pyodide building on my ARM machine natively, **the rebuild process is at least faster, only taking 3 minutes** rather than 20 minutes (when building on ARM macOS with an x86 Linux Docker container):

```
$ time ( make -C cpython clean && make )
real	3m5.069s
user	2m12.576s
sys	0m37.067s
$ echo $?
0
```

**Edit:** This rebuild process was fast enough for me to remove all the optional modules I could, while keeping the resulting python binary working.
